File: Data Structures & Algorithms/permutations/submission-0.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []

        def dfs(path, available):
            if not available:  # no numbers left to pick
                res.append(path[:])
                return

            for num in list(available):  # iterate over a snapshot
                available.remove(num)
                path.append(num)

                dfs(path, available)

                # backtrack
                path.pop()
                available.add(num)

        dfs([], set(nums))
        return res

File: Data Structures & Algorithms/two-integer-sum-ii/submission-0.java
class Solution {
    public int[] twoSum(int[] numbers, int target) {
        int[] solution = new int[2];
        int left = 0;
        int right = numbers.length-1;

        while (left < right) {
            if (numbers[left] + numbers[right] == target) {
                solution[0] = left+1;
                solution[1] = right+1;
                break;
            } else if (numbers[left] + numbers[right] < target) {
                left++;
            } else {
                right--;
            }
        }

        return solution;
    }
}
